File: solizps/cf_oii/G1/gen.cpp
#include "testlib.h"
#include <bits/stdc++.h>
using namespace std;

#define nl "\n"
#define nf endl
#define ll long long
#define pb push_back
#define _ << ' ' <<

#define INF (ll)1e18
#define mod 998244353
#define maxn 200010

int main(int argc, char* argv[]) {
    registerGen(argc, argv, 1);

    ll t = atoi(argv[1]);
    ll n_min = atoi(argv[2]);
    ll n_max = atoi(argv[3]);
    ll type = atoi(argv[4]);

    auto gauss = [&](ll x) {
        return (x * (x + 1)) / 2;
    };

    cout << t << nl;
    for (ll t_id = 1; t_id <= t; t_id++) {
        ll n = rnd.next(n_min, n_max);
        vector<int> l(n, 0), r(n, 0);
        const int MAXT = 1000000000; // Compress limit

        for (int i = 0; i < n; ++i) {
            l[i] = rnd.next(1, MAXT);
            r[i] = rnd.next(1, MAXT);
            if (r[i] < l[i]) swap(l[i], r[i]);
        }

        if (type == 1) {  // max_k(m) assume Theta(sqrt(n)) valori distinti
            int t = 1;
            while (gauss(t) <= n) {
                int mid = 30 * gauss(t); // Punto di intersezione del blocco da t
                for (int i = 0; i < t; ++i) {
                    l[gauss(t - 1) + i] = rnd.next(mid - 10 * t, mid);
                    r[gauss(t - 1) + i] = rnd.next(mid, mid + 10 * t);
                }
                ++t;
            }
        } else if (type == 2) {  // M grande
            for (int i = 0; i < rnd.next((ll)(n * 0.9), n); ++i) {
                l[i] = rnd.next((ll)1, n);
                r[i] = n + rnd.next((ll)1, n);
            }
        } else if (type == 3) {
            for (int i = 0; i < rnd.next((ll)(n * 0.9), n); ++i) {
                l[i] = 2 * i;
                r[i] = 2 * i + 1;
            }
        } else if (type == 4) {
            int m = rnd.next((ll)1, n);
            int k = n / m;
            for (int i = 0; i < k; ++i) {
                int mid = 30 * m * (i + 1);
                for (int j = 0; j < m; ++j) {
                    l[i * m + j] = rnd.next(mid - 10 * m, mid);
                    r[i * m + j] = rnd.next(mid, mid + 10 * m);
                }
            }
        } else if (type == 5) {
            ll maxc = atoi(argv[5]);
            ll maxl = atoi(argv[6]);
            assert(maxl <= maxc);
            for (int i = 0; i < n; ++i) {
                ll len = rnd.next((ll)0, maxl);
                l[i] = rnd.next((ll)1, maxc - len);
                r[i] = l[i] + len;
            }
        }

        // Coordinate compression
        vector<int> z = l;
        z.insert(z.end(), r.begin(), r.end());
        sort(z.begin(), z.end());
        z.erase(unique(z.begin(), z.end()), z.end());

        vector<int> targets(z.size(), 0);
        for (int i = 0; i < z.size(); i++) {
            targets[i] = rnd.next((ll)1, 2 * n - (ll)z.size() + 1);
        }
        sort(targets.begin(), targets.end());
        for (int i = 0; i < z.size(); i++) targets[i] += i;

        map<int, int> mp;
        for (int i = 0; i < z.size(); ++i) {
            mp[z[i]] = targets[i];
        }

        for (int i = 0; i < n; ++i) {
            l[i] = mp[l[i]];
            r[i] = mp[r[i]];
        }

        // Output
        cout << n << nl;
        for (int i = 0; i < n; ++i) {
            cout << l[i];
            if (i == n - 1) cout << nl;
            else cout << ' ';
        }
        for (int i = 0; i < n; ++i) {
            cout << r[i];
            if (i == n - 1) cout << nl;
            else cout << ' ';
        }
    }
}

File: solizps/cf_oii/G1/gen.py
#!/usr/bin/env python3

from limiti import *

from sys import argv, exit, stderr
import os
import math
from random import choice, sample, shuffle, seed, random, randint, choices

"""
Possibli lamerate a cui rendere la vita difficile:
    * assumere M piccolo
    * assumere M in un range
    * assumere che i testcase siano fatti male e, quando M è artificialmente,
      grande, ci sono pochi intervalli distinti
"""

def run(N, st, gen_type):
    for k, v in subtask[st].items():
        globals()[k] = v

    l, r = [], []

    MAXT = 10**9 # poi comprimo

    if Keq2:
        # fissa tre punti x1 < x2 < x3. x1, x2 saranno di intersezione
        # gli intervalli di sinistra sono contenuti in [1, x3-1], gli altri
        # in [x3+1, MAXT]. La risposta consiste di M intervalli contenuti in
        # [1, x3-1] e altrettanti contenuti in [x3+1, MAXT]. I restanti N - 2*M
        # intervalli si possono piazzare a caso senza alterare la risposta, a
        # patto che M > N/3.
        x2 = randint(1, MAXT)
        while x2 == 1 or x2 == MAXT:
            x2 = randint(1, MAXT)
        x1 = randint(1, x2 - 1)
        x3 = randint(x2 + 1, MAXT)

        M = randint(N//3 + 1, N // 2);
        for i in range(M):
            l.append(randint(1, x1))
            r.append(randint(x1, x2 - 1))
        for i in range(M):
            l.append(randint(x2 + 1, x3))
            r.append(randint(x3, MAXT))
        for i in range(N - 2 * M):
            l.append(randint(1, MAXT))
            r.append(randint(1, MAXT))
            if r[-1] < l[-1]:
                l[-1], r[-1] = r[-1], l[-1]
    else:
        l = [randint(1, MAXT) for x in range(N)]
        r = [randint(1, MAXT) for x in range(N)]
        for i in range(N):
            if r[i] < l[i]:
                l[i], r[i] = r[i], l[i]
        if gen_type == 1:           # max_k(m) assume Theta(sqrt(n)) valori distinti
            Gauss = lambda x : x * (x + 1) // 2
            t = 1
            while Gauss(t) <= N:
                mid = 30 * Gauss(t) # punto di intersezione del blocco da t
                for i in range(t):
                    l[Gauss(t-1) + i] = randint(mid - 10*t, mid)
                    r[Gauss(t-1) + i] = randint(mid, mid + 10*t)
                t += 1
        elif gen_type == 2:         # M grande
            for i in range(randint(int(N*0.9), N)):
                l[i] = randint(1, N)
                r[i] = N + randint(1, N)
        elif gen_type == 3:
            for i in range(randint(int(N*0.9), N)):
                l[i] = 2 * i
                r[i] = 2 * i + 1
        elif gen_type == 4:
            M = randint(1, N)
            K = N // M;
            for i in range(K):
                mid = 30 * M * (i + 1)
                for j in range(M):
                    l[i * K + j] = randint(mid - 10 * M, mid)
                    r[i * K + j] = randint(mid, mid + 10 * M)

    # coordinate compression
    z = l + r;
    zz = sorted(set(z))
    mp = {value: index for index, value in enumerate(zz)}
    for i in range(N):
        l[i] = 1 + mp[l[i]]
        r[i] = 1 + mp[r[i]]

    print(N)
    print(' '.join(str(x) for x in l))
    print(' '.join(str(x) for x in r))

if __name__ == "__main__":
    *args, S = map(int, argv[1:])

    seed(S)
    run(*args)
